### DISCUSSION TITLE
is_connected() . Trying to set up an alert.

### DISCUSSION THREAD

**Discussion by mydroidbox** [pschatzmann/ESP32-A2DP#387] [Q&A] [answered]
Hey Phil. "Here I go again on my own," White Tiger.

Trying to get an alert when bluetooth is connected.
Getting not declared in this scope. 

/**
   @file base-player-a2dp.ino
   @author Phil Schatzmann
   @brief Sketch which uses the A2DP callback to provide data from the AudioPlayer via a Queue

   @version 0.1
   @date 2022-12-04

   @copyright Copyright (c) 2022

*/

#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"
#include "AudioLibs/AudioSourceSDFAT.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "AudioLibs/AudioKit.h"

const int led = 5;

int buffer_count = 30;
int buffer_size = 512;
const char *startFilePath = "/";
const char* ext = "mp3";
AudioSourceSDFAT source(startFilePath, ext, PIN_AUDIO_KIT_SD_CARD_CS);
MP3DecoderHelix decoder;
//Setup of synchronized buffer
SynchronizedNBuffer<uint8_t> buffer(buffer_size, buffer_count, portMAX_DELAY, 10);
QueueStream<uint8_t> out(buffer); // convert Buffer to Stream
AudioPlayer player(source, out, decoder);
BluetoothA2DPSource a2dp;

  // Provide data to A2DP
  int32_t get_data(uint8_t *data, int32_t bytes) {
  size_t result_bytes = buffer.readArray(data, bytes);
  //LOGI("get_data_channels %d -> %d of (%d)", bytes, result_bytes , buffer.available());
  return result_bytes;
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  pinMode(led, OUTPUT);

  // sd_active is setting up SPI with the right SD pins by calling
  SPI.begin(PIN_AUDIO_KIT_SD_CARD_CLK, PIN_AUDIO_KIT_SD_CARD_MISO, PIN_AUDIO_KIT_SD_CARD_MOSI, PIN_AUDIO_KIT_SD_CARD_CS);
  
  // start QueueStream
  out.begin();

  // setup player
  player.setDelayIfOutputFull(0);
  player.setVolume(0.6);
  player.begin();

  // fill buffer with some data
  player.getStreamCopy().copyN(5);

  // start a2dp source
  Serial.println("starting A2DP...");
  a2dp.start_raw("HW2-915",  get_data);
  Serial.println("Started!");

}

void loop() {

  if (is_connected()) {
    digitalWrite(led, HIGH);
  }

  player.copy();

}
`

**Comment by pschatzmann** [comment, marked as answer]
A2DP is an object oriented library! You forgot to add the object which should execute the method: a2dp.is_connected()

**Reply by mydroidbox** [reply]
Ahh...got you. Thanks Phil. One other quick
question in this sketch.
trying to add alternative bluetooth speakers.
but when I run code below I get the error:
'a2dp_source' was not declared in this scope

I also tried a2dp.start_raw but got:
no matching function for call to 'BluetoothA2DPSource::start_raw(std::vector<char*>&, int32_t (&)(uint8_t*, int32_t))'


`
/**
@file base-player-a2dp.ino
@author Phil Schatzmann
@brief Sketch which uses the A2DP callback to provide data from the AudioPlayer via a Queue

@Version 0.1
@Date 2022-12-04

@copyright Copyright (c) 2022

*/

#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"
#include "AudioLibs/AudioSourceSDFAT.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "AudioLibs/AudioKit.h"

const int led = 5;

int buffer_count = 30;
int buffer_size = 512;
const char startFilePath = "/";
const char ext = "mp3";
AudioSourceSDFAT source(startFilePath, ext, PIN_AUDIO_KIT_SD_CARD_CS);
MP3DecoderHelix decoder;
//Setup of synchronized buffer
SynchronizedNBuffer<uint8_t> buffer(buffer_size, buffer_count, portMAX_DELAY, 10);
QueueStream<uint8_t> out(buffer); // convert Buffer to Stream
AudioPlayer player(source, out, decoder);
BluetoothA2DPSource a2dp;

// Provide data to A2DP
int32_t get_data(uint8_t *data, int32_t bytes) {
size_t result_bytes = buffer.readArray(data, bytes);
//LOGI("get_data_channels %d -> %d of (%d)", bytes, result_bytes , buffer.available());
return result_bytes;
}

void setup() {
Serial.begin(115200);
AudioLogger::instance().begin(Serial, AudioLogger::Warning);

static std::vector<char*> bt_names = {"HW2-915","WP522"};

pinMode(led, OUTPUT);

// sd_active is setting up SPI with the right SD pins by calling
SPI.begin(PIN_AUDIO_KIT_SD_CARD_CLK, PIN_AUDIO_KIT_SD_CARD_MISO, PIN_AUDIO_KIT_SD_CARD_MOSI, PIN_AUDIO_KIT_SD_CARD_CS);

// start QueueStream
out.begin();

// setup player
player.setDelayIfOutputFull(0);
player.setVolume(0.6);
player.begin();

// fill buffer with some data
player.getStreamCopy().copyN(5);

// start a2dp source
Serial.println("starting A2DP...");
//a2dp.start_raw("HW2-915", get_data);
a2dp_source.start(bt_names, get_data); 
Serial.println("Started!");

}

void loop() {

if (a2dp.is_connected()){
digitalWrite(led, HIGH);
}

player.copy();

}
`

**Comment by pschatzmann** [comment]
Check the documentation and you should see your error: https://pschatzmann.github.io/ESP32-A2DP/html/class_bluetooth_a2_d_p_source.html

**Reply by mydroidbox** [reply]
Not seeing it. I see where both "start" and "start_raw" support multiple bluetooth names.
Can't get either to work.

**Reply by pschatzmann** [reply]
It requires a std::vector<const char*> and not std::vector<char*> !

**Reply by mydroidbox** [reply]
I got to tell you, that is pretty freaking cool to have more than one option for a speaker.
Appreciated.
